File: client/android/app/src/main/java/org/tinymux/titan/data/SessionLogger.kt
package org.tinymux.titan.data

import android.content.Context
import java.io.File
import java.io.FileWriter
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class SessionLogger(private val context: Context) {
    private var writer: FileWriter? = null
    private var logFile: File? = null
    var active = false; private set

    val logDir: File get() = File(context.filesDir, "logs").also { it.mkdirs() }

    fun start(worldName: String, filename: String? = null): File {
        stop()
        val safeName = worldName.replace(Regex("[^a-zA-Z0-9._-]"), "_")
        val ts = SimpleDateFormat("yyyyMMdd_HHmmss", Locale.US).format(Date())
        val name = filename ?: "${safeName}_$ts.log"
        val file = File(logDir, name)
        writer = FileWriter(file, true)
        logFile = file
        active = true
        writeLine("--- Log started: ${Date()} ---")
        return file
    }

    fun stop() {
        if (!active) return
        writeLine("--- Log ended: ${Date()} ---")
        try { writer?.close() } catch (_: Exception) {}
        writer = null
        logFile = null
        active = false
    }

    fun writeLine(line: String) {
        if (!active) return
        try {
            writer?.write(line)
            writer?.write("\n")
            writer?.flush()
        } catch (_: Exception) {}
    }

    fun currentFile(): File? = logFile
}

File: client/ios/Titan/Service/SessionLogger.swift
import Foundation

// MARK: - Session Logger

class SessionLogger {
    private var fileHandle: FileHandle?
    private var logFile: URL?
    private(set) var active = false

    var logDir: URL {
        let docs = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!
        let dir = docs.appendingPathComponent("logs", isDirectory: true)
        try? FileManager.default.createDirectory(at: dir, withIntermediateDirectories: true)
        return dir
    }

    func start(worldName: String, filename: String? = nil) -> URL {
        stop()
        let safe = worldName.replacingOccurrences(of: "[^a-zA-Z0-9._-]", with: "_",
                                                   options: .regularExpression)
        let formatter = DateFormatter()
        formatter.dateFormat = "yyyyMMdd_HHmmss"
        let ts = formatter.string(from: Date())
        let name = filename ?? "\(safe)_\(ts).log"
        let file = logDir.appendingPathComponent(name)
        FileManager.default.createFile(atPath: file.path, contents: nil)
        fileHandle = try? FileHandle(forWritingTo: file)
        fileHandle?.seekToEndOfFile()
        logFile = file
        active = true
        writeLine("--- Log started: \(Date()) ---")
        return file
    }

    func stop() {
        guard active else { return }
        writeLine("--- Log ended: \(Date()) ---")
        try? fileHandle?.close()
        fileHandle = nil
        logFile = nil
        active = false
    }

    func writeLine(_ line: String) {
        guard active, let data = "\(line)\n".data(using: .utf8) else { return }
        fileHandle?.write(data)
    }

    func currentFile() -> URL? { logFile }
}
